From 9d147c916c5987e52ca104af6ef7ad24bfd5b02e Mon Sep 17 00:00:00 2001
From: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Date: Thu, 23 Aug 2018 18:20:02 +0900
Subject: [PATCH 294/909] drm: rcar-du: Fix DPLL clock divider judgement

commit 75327296b402e5bf3b457fa584174a31194304b2 from
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-bsp.git

There is case that Display driver selects dot clock less than
request dot clock in DPLL calculation. The driver can output more
stable by selecting a larger clock than the request clock in high
resolution. This patch corrects it.

Signed-off-by: Koji Matsuoka <koji.matsuoka.xm@renesas.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 drivers/gpu/drm/rcar-du/rcar_du_crtc.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
index 309c9b34e65c..c6b813415104 100644
--- a/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
+++ b/drivers/gpu/drm/rcar-du/rcar_du_crtc.c
@@ -124,6 +124,10 @@ static void rcar_du_dpll_divider(struct rcar_du_crtc *rcrtc,
 	unsigned int fdpll;
 	unsigned int m;
 	unsigned int n;
+	bool clk_high = false;
+
+	if (target > 148500000)
+		clk_high = true;
 
 	/*
 	 *   fin                                 fvco        fout       fclkout
@@ -176,6 +180,9 @@ static void rcar_du_dpll_divider(struct rcar_du_crtc *rcrtc,
 				if (output >= 400 * 1000 * 1000)
 					continue;
 
+				if (clk_high && output < target)
+					continue;
+
 				diff = abs((long)output - (long)target);
 				if (best_diff > diff) {
 					best_diff = diff;
-- 
2.17.1

